In [5]:
import pandas as pd
import time

start = time.time()

X_train = pd.read_csv("../../data/processed/X_train.csv")

print("X_train shape:", X_train.shape)
print("Time:", round(time.time() - start, 2), "seconds")

X_train shape: (24000, 34)
Time: 0.25 seconds


In [7]:
start = time.time()

X_test = pd.read_csv("../../data/processed/X_test.csv")

print("X_test shape:", X_test.shape)
print("Time:", round(time.time() - start, 2), "seconds")

X_test shape: (6000, 34)
Time: 0.06 seconds


In [8]:
start = time.time()

y_train = pd.read_csv("../../data/processed/y_train.csv").squeeze()

print("y_train shape:", y_train.shape)
print("Time:", round(time.time() - start, 2), "seconds")

y_train shape: (24000,)
Time: 0.02 seconds


In [9]:
start = time.time()

y_test = pd.read_csv("../../data/processed/y_test.csv").squeeze()

print("y_test shape:", y_test.shape)
print("Time:", round(time.time() - start, 2), "seconds")

y_test shape: (6000,)
Time: 0.01 seconds


In [11]:
from pathlib import Path

processed_dir = Path("../../data/processed")

for file in processed_dir.iterdir():
    print(file.name, round(file.stat().st_size / (1024 * 1024), 2), "MB")

X_test.csv 1.21 MB
X_train.csv 4.84 MB
y_test.csv 0.02 MB
y_train.csv 0.07 MB


In [12]:
print("X_train:", X_train.shape)
print("X_test :", X_test.shape)
print("y_train:", y_train.shape)
print("y_test :", y_test.shape)

print("\nData types:")
print(X_train.dtypes)

print("\nMissing values:")
print("X_train:", X_train.isnull().sum().sum())
print("X_test :", X_test.isnull().sum().sum())

X_train: (24000, 34)
X_test : (6000, 34)
y_train: (24000,)
y_test : (6000,)

Data types:
appeared_year               float64
appeared_month              float64
hist_mean                   float64
hist_std                    float64
hist_min                    float64
hist_max                    float64
hist_sum                    float64
entropy_mean                float64
entropy_std                 float64
entropy_min                 float64
entropy_max                 float64
strings_numstrings          float64
strings_avlength            float64
strings_printables          float64
strings_entropy             float64
strings_paths               float64
strings_urls                float64
strings_registry            float64
strings_MZ                  float64
general_size                float64
general_vsize               float64
general_has_debug           float64
general_exports             float64
general_imports             float64
general_has_relocations     float64
general_has

In [13]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

lr_model = LogisticRegression(
    max_iter=1000,
    random_state=42
)

lr_model.fit(X_train, y_train)

y_pred = lr_model.predict(X_test)

lr_accuracy = accuracy_score(y_test, y_pred)
lr_precision = precision_score(y_test, y_pred)
lr_recall = recall_score(y_test, y_pred)
lr_f1 = f1_score(y_test, y_pred)

print("Logistic Regression")
print("-------------------")
print("Accuracy :", round(lr_accuracy, 4))
print("Precision:", round(lr_precision, 4))
print("Recall   :", round(lr_recall, 4))
print("F1 Score :", round(lr_f1, 4))

Logistic Regression
-------------------
Accuracy : 0.681
Precision: 0.83
Recall   : 0.4987
F1 Score : 0.6231


c:\Users\DELL\Desktop\Springboard\Malware-Classification-System\.venv\Lib\site-packages\sklearn\linear_model\_logistic.py:599: ConvergenceWarning: lbfgs failed to converge after 1000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=1000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [14]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

lr_model = LogisticRegression(
    max_iter=2000,
    random_state=42
)

lr_model.fit(X_train_scaled, y_train)

y_pred_lr = lr_model.predict(X_test_scaled)

lr_accuracy = accuracy_score(y_test, y_pred_lr)
lr_precision = precision_score(y_test, y_pred_lr)
lr_recall = recall_score(y_test, y_pred_lr)
lr_f1 = f1_score(y_test, y_pred_lr)

print("Logistic Regression (Scaled)")
print("---------------------------")
print("Accuracy :", round(lr_accuracy, 4))
print("Precision:", round(lr_precision, 4))
print("Recall   :", round(lr_recall, 4))
print("F1 Score :", round(lr_f1, 4))

Logistic Regression (Scaled)
---------------------------
Accuracy : 0.7733
Precision: 0.8034
Recall   : 0.7563
F1 Score : 0.7791


In [15]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

extra_trees = ExtraTreesClassifier(
    n_estimators=100,
    random_state=42,
    n_jobs=-1
)

extra_trees.fit(X_train, y_train)

y_pred_et = extra_trees.predict(X_test)

et_accuracy = accuracy_score(y_test, y_pred_et)
et_precision = precision_score(y_test, y_pred_et)
et_recall = recall_score(y_test, y_pred_et)
et_f1 = f1_score(y_test, y_pred_et)

print("Extra Trees")
print("-----------")
print("Accuracy :", round(et_accuracy, 4))
print("Precision:", round(et_precision, 4))
print("Recall   :", round(et_recall, 4))
print("F1 Score :", round(et_f1, 4))

Extra Trees
-----------
Accuracy : 0.9315
Precision: 0.9498
Recall   : 0.919
F1 Score : 0.9341


In [16]:
import xgboost as xgb
from xgboost import XGBClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

xgb_model = XGBClassifier(
    n_estimators=100,
    max_depth=6,
    learning_rate=0.1,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42,
    n_jobs=-1,
    eval_metric="logloss"
)

xgb_model.fit(X_train, y_train)

y_pred_xgb = xgb_model.predict(X_test)

xgb_accuracy = accuracy_score(y_test, y_pred_xgb)
xgb_precision = precision_score(y_test, y_pred_xgb)
xgb_recall = recall_score(y_test, y_pred_xgb)
xgb_f1 = f1_score(y_test, y_pred_xgb)

print("XGBoost")
print("-------")
print("Accuracy :", round(xgb_accuracy, 4))
print("Precision:", round(xgb_precision, 4))
print("Recall   :", round(xgb_recall, 4))
print("F1 Score :", round(xgb_f1, 4))

XGBoost
-------
Accuracy : 0.9128
Precision: 0.9247
Recall   : 0.9092
F1 Score : 0.9169


In [17]:
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

hgb_model = HistGradientBoostingClassifier(
    max_iter=100,
    learning_rate=0.1,
    max_leaf_nodes=31,
    random_state=42
)

hgb_model.fit(X_train, y_train)

y_pred_hgb = hgb_model.predict(X_test)

hgb_accuracy = accuracy_score(y_test, y_pred_hgb)
hgb_precision = precision_score(y_test, y_pred_hgb)
hgb_recall = recall_score(y_test, y_pred_hgb)
hgb_f1 = f1_score(y_test, y_pred_hgb)

print("HistGradientBoosting")
print("--------------------")
print("Accuracy :", round(hgb_accuracy, 4))
print("Precision:", round(hgb_precision, 4))
print("Recall   :", round(hgb_recall, 4))
print("F1 Score :", round(hgb_f1, 4))

HistGradientBoosting
--------------------
Accuracy : 0.9157
Precision: 0.9306
Recall   : 0.9083
F1 Score : 0.9193


In [18]:
import lightgbm
from lightgbm import LGBMClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)

lgbm_model = LGBMClassifier(
    n_estimators=100,
    learning_rate=0.1,
    max_depth=-1,
    num_leaves=31,
    random_state=42,
    n_jobs=-1,
    verbosity=-1
)

lgbm_model.fit(X_train, y_train)

y_pred_lgbm = lgbm_model.predict(X_test)

lgbm_accuracy = accuracy_score(y_test, y_pred_lgbm)
lgbm_precision = precision_score(y_test, y_pred_lgbm)
lgbm_recall = recall_score(y_test, y_pred_lgbm)
lgbm_f1 = f1_score(y_test, y_pred_lgbm)

print("LightGBM")
print("--------")
print("Accuracy :", round(lgbm_accuracy, 4))
print("Precision:", round(lgbm_precision, 4))
print("Recall   :", round(lgbm_recall, 4))
print("F1 Score :", round(lgbm_f1, 4))

LightGBM
--------
Accuracy : 0.9152
Precision: 0.9294
Recall   : 0.9086
F1 Score : 0.9189


In [19]:
import pandas as pd

baseline_results = pd.DataFrame({
    "Model": [
        "Random Forest",
        "Extra Trees",
        "Logistic Regression",
        "HistGradientBoosting",
        "XGBoost",
        "LightGBM"
    ],
    
    "Accuracy": [
        0.9350,
        et_accuracy,
        lr_accuracy,
        hgb_accuracy,
        xgb_accuracy,
        lgbm_accuracy
    ],
    
    "Precision": [
        0.9624,
        et_precision,
        lr_precision,
        hgb_precision,
        xgb_precision,
        lgbm_precision
    ],
    
    "Recall": [
        0.9127,
        et_recall,
        lr_recall,
        hgb_recall,
        xgb_recall,
        lgbm_recall
    ],
    
    "F1 Score": [
        0.9369,
        et_f1,
        lr_f1,
        hgb_f1,
        xgb_f1,
        lgbm_f1
    ]
})

baseline_results

,Model,Accuracy,Precision,Recall,F1 Score
0,Random Forest,0.935000,0.962400,0.912700,0.936900
1,Extra Trees,0.931500,0.949821,0.918979,0.934145
2,Logistic Regression,0.773333,0.803416,0.756305,0.779149
3,HistGradientBoosting,0.915667,0.930556,0.908260,0.919272
4,XGBoost,0.912833,0.924655,0.909206,0.916865
5,LightGBM,0.915167,0.929378,0.908575,0.918859


In [20]:
display_table = baseline_results.copy()

for column in ["Accuracy", "Precision", "Recall", "F1 Score"]:
    display_table[column] = (
        display_table[column] * 100
    ).round(2).astype(str) + "%"

display_table

,Model,Accuracy,Precision,Recall,F1 Score
0,Random Forest,93.5%,96.24%,91.27%,93.69%
1,Extra Trees,93.15%,94.98%,91.9%,93.41%
2,Logistic Regression,77.33%,80.34%,75.63%,77.91%
3,HistGradientBoosting,91.57%,93.06%,90.83%,91.93%
4,XGBoost,91.28%,92.47%,90.92%,91.69%
5,LightGBM,91.52%,92.94%,90.86%,91.89%


In [21]:
best_accuracy = baseline_results.loc[
    baseline_results["Accuracy"].idxmax()
]

best_recall = baseline_results.loc[
    baseline_results["Recall"].idxmax()
]

best_f1 = baseline_results.loc[
    baseline_results["F1 Score"].idxmax()
]

print("Best Accuracy:")
print(
    best_accuracy["Model"],
    f"({best_accuracy['Accuracy']:.2%})"
)

print("\nBest Malware Recall:")
print(
    best_recall["Model"],
    f"({best_recall['Recall']:.2%})"
)

print("\nBest F1 Score:")
print(
    best_f1["Model"],
    f"({best_f1['F1 Score']:.2%})"
)

Best Accuracy:
Random Forest (93.50%)

Best Malware Recall:
Extra Trees (91.90%)

Best F1 Score:
Random Forest (93.69%)


In [23]:
baseline_results.to_csv(
    "../../reports/baseline_model_comparison.csv",
    index=False
)

print("Baseline comparison saved successfully!")

Baseline comparison saved successfully!


# Final Baseline Model Comparison

Six classification algorithms were evaluated using the same training and test datasets and the same 34 engineered features. The models were compared using accuracy, precision, recall, and F1 score.

The Random Forest model achieved the highest overall F1 score (93.69%) and accuracy (93.50%), while Extra Trees achieved the highest malware recall (91.90%).